drivers: ieee802154: replace NET_CONFIG_IEEE802154_DEV_NAME with zephyr,ieee802154 choice

gmarull opened the pull request:
Thanks to IEEE 802.15.4 drivers conversion to DT, we can finally remove `NET_CONFIG_IEEE802154_DEV_NAME` which forced usage of `device_get_binding()` when working with IEEE 802.15.4 devices. Instead, the chosen IEEE 802.15.4 device is specified in Devicetree using `zephyr,ieee802154`.

All boards and samples have been updated.

gmarull committed:
tests: drivers: build_all: ieee802154: fix test case name

CC13XX/CC26XX test case had wrong name, fix it.

Signed-off-by: Gerard Marull-Paretas <gerard.marull@nordicsemi.no>

gmarull committed:
modules: openthread: use zephyr,ieee802154 choice

Use chosen ieee802154 device from DT since
CONFIG_NET_CONFIG_IEEE802154_DEV_NAME is being phased out.

Signed-off-by: Gerard Marull-Paretas <gerard.marull@nordicsemi.no>

gmarull committed:
tests: subsys: openthread: refactor test

Use a test compatible so that test can create a DT-based device and
provide a valid choice when building the test.

Signed-off-by: Gerard Marull-Paretas <gerard.marull@nordicsemi.no>

gmarull committed:
net: lib: config: ieee802154: use zephyr,ieee802154

Use DT choice zephyr,ieee802154 as CONFIG_NET_CONFIG_IEEE802154_DEV_NAME
is being phased out.

Signed-off-by: Gerard Marull-Paretas <gerard.marull@nordicsemi.no>

gmarull committed:
samples: net: wpanusb: use zephyr,ieee802154

Use the choisen zephyr,ieee802154 from DT as
CONFIG_NET_CONFIG_IEEE802154_DEV_NAME is being phased out.

Signed-off-by: Gerard Marull-Paretas <gerard.marull@nordicsemi.no>

gmarull committed:
samples: net: wpan_serial: use zephyr,ieee802154

Use the choisen zephyr,ieee802154 from DT as
CONFIG_NET_CONFIG_IEEE802154_DEV_NAME is being phased out.

Signed-off-by: Gerard Marull-Paretas <gerard.marull@nordicsemi.no>

gmarull committed:
samples: net: openthread: coprocessor: remove Kconfig IEEE802154 choice

Remove Kconfig choice for IEEE802154. In this case, there's no in-tree
board, so there's no overlay or board to be updated.

Signed-off-by: Gerard Marull-Paretas <gerard.marull@nordicsemi.no>

gmarull committed:
net: lib: config: remove NET_CONFIG_IEEE802154_DEV_NAME

Remove NET_CONFIG_IEEE802154_DEV_NAME in favor of DT based choice using
zephyr,ieee802154.

Signed-off-by: Gerard Marull-Paretas <gerard.marull@nordicsemi.no>

gmarull committed:
doc: release-notes: add IEEE 802.15.4 changes

- All device drivers are now DT based
- New choice has been added for the subsystem: zephyr,ieee802154.

Signed-off-by: Gerard Marull-Paretas <gerard.marull@nordicsemi.no>

gmarull committed:
tests: net: ieee802154: use native_posix

All IEEE 802.15.4 tests use internal fake devices, so use native_posix
platform to run them as they are not meant to test IEEE 802.15.4 on
hardware.

Signed-off-by: Gerard Marull-Paretas <gerard.marull@nordicsemi.no>

gmarull committed:
samples: net: wpan(usb|serial): filter based on zephyr,ieee802154

Rely on zephyr,ieee802154 choice being enabled for the samples.

Signed-off-by: Gerard Marull-Paretas <gerard.marull@nordicsemi.no>

gmarull committed:
boards: enable IEEE 802.15.4 devices and add choice

Enable the IEEE 802.15.4 radio node on all boards that listed
'ieee802154' in the supported field.

Signed-off-by: Gerard Marull-Paretas <gerard.marull@nordicsemi.no>

gmarull committed:
boards: remove unused tag from supported list

All IEEE 802.15.4 tests/samples use DT-based filter now, so the
ieee802154 tag can be deleted.

Signed-off-by: Gerard Marull-Paretas <gerard.marull@nordicsemi.no>

gmarull committed:
tests: drivers: build_all: ieee802154: refactor test config

- Use a native_posix specific overlay for non-SoC devices
- Remove redundant per-driver Kconfig selections

Signed-off-by: Gerard Marull-Paretas <gerard.marull@nordicsemi.no>

rerickson1 reviewed (dismissed):
All Laird Connectivity boards look good!

galak reviewed (changes requested):
Can you break up this PR.  There's too much going on to easily review it.

gmarull commented:
> Can you break up this PR. There's too much going on to easily review it.

Changes are fairly trivial when reviewed on a per-commit basis. Changes are grouped here to actually minimize changes and avoid unnecessary intermediate stages (e.g. keep labels).

gmarull committed:
samples: net: sockets: echo_(client|server): remove IEEE802154_DEV_NAME

Remove usage of CONFIG_NET_CONFIG_IEEE802154_DEV_NAME as it now works
with DT based choice.

Signed-off-by: Gerard Marull-Paretas <gerard.marull@nordicsemi.no>

gmarull committed:
doc: build: dts: api: add zephyr,ieee802154

Add an entry explaining the purpose of zephyr,ieee802154

Signed-off-by: Gerard Marull-Paretas <gerard.marull@nordicsemi.no>

rlubos commented on drivers/ieee802154/Kconfig.nrf5:
```diff
@@ -11,12 +11,6 @@ menuconfig IEEE802154_NRF5
 
 if IEEE802154_NRF5
 
-config IEEE802154_NRF5_DRV_NAME
```
So how can be a device name specified now, I've tried to check what is the name actually assigned with those changes, but it seems to be an empty string.

I'm asking, as the network device name is particularly important in terms of networking, as it also identifies individual network interfaces, w/o a name application is not able to use `SO_BINDTODEVICE` socket option for example.

gmarull replied:
I guess one can always use `DEVICE_DT_NAME_GET`. @galak any comments here? it may be relevant for the `label` removals.

rlubos reviewed (commented):
Generally, looks good, however, the device naming is not clear to me.

gmarull commented:
> @gmarull - any thoughts about devices that provide > 1 ieee802154 radios? e.g. cc13xx can have simultaneous 2.4 GHz and Sub-GHz radios active, each with a network configuration.
> 
> I like the DT approach. However, in general, I think we need to have dt bindings to configure multiple network interfaces as well (cc @rlubos). That way we can use `DT_INST_FOREACH_STATUS_OKAY()` for all network interfaces. #29750

That's a good question because there's one single choice. However, I guess it was this way now as well because we only had a single Kconfig option. Any ideas?

cfriedt commented:
> That's a good question because there's one single choice. However, I guess it was this way now as well because we only had a single Kconfig option. Any ideas?

Network configuration in DT would solve all of these problems, no? I wonder if we could expand on #29750 an API meeting. DT is really the ideal place for this because it gives us multiple-nodes as variable-length lists of things like static / multiple IP addresses, DNS, etc. I would be happy to chat a bit with you and @rlubos for some details.

nandojve commented:
I would like see this PR move forward somehow. The `IEEE802154_<MANUFACTURER>_DRV_NAME` have a good purpose in past but today we need to move to a better approach. I'm not sure if we can solve all the problems right now and multiple interfaces are desired for sure, mostly when there is necessity to use dual band like 900MHz and 2.4GHz and those could be from different manufacturers, inclusive. There is AT86RF215 pending on my todo list and it supports 900MHz and 2.4GHz too.

cfriedt commented:
@gmarull @nandojve - I agree. I think this can move forward. Will be handling DT bindings with a separate RFC

galak commented on drivers/ieee802154/Kconfig.b91:
```diff
@@ -5,7 +5,8 @@
 	bool "Telink B91 RF driver"
-	depends on NETWORKING
+	default y
+	depends on NETWORKING && DT_HAS_TELINK_B91_ZB_ENABLED
```
seems like NETWORKING should be up a level and not at driver level??

gmarull replied:
moved one level up

galak commented on drivers/ieee802154/Kconfig.cc13xx_cc26xx:
```diff
@@ -60,16 +56,12 @@ endif # IEEE802154_CC13XX_CC26XX
 
 menuconfig IEEE802154_CC13XX_CC26XX_SUB_GHZ
```
Do we know what's going on here between SUB_GHZ and non-SUB_GHZ?  Are these really 2 different devices?

gmarull replied:
apparently they are 2 distinct drivers

galak reviewed (changes requested):
Made some comments, but still probably have others.  I think its better to split up this PR per driver, and than all the other changes after.

gmarull commented on dts/arm/nxp/nxp_kw41z.dtsi:
```diff
@@ -258,6 +258,17 @@
+			reg = <0x4005d000 0x1000>;
+			status = "disabled";
+
+			ieee802154: ieee802154 {
```
left ieee802154 node only with the 0x4005d000 reg (driver actually doesn't care about what's in DT)

galak reviewed (changes requested):
What happen to dts update for boards/arm/nrf52840_blip?

galak reviewed (commented):
Can you remove `label` usage from here as part of this PR:

```
dts/arm/nordic/nrf52811.dtsi:				label = "IEEE802154_nrf5";
dts/arm/nordic/nrf52833.dtsi:				label = "IEEE802154_nrf5";
dts/arm/nordic/nrf52840.dtsi:				label = "IEEE802154_nrf5";
dts/arm/nordic/nrf5340_cpunet.dtsi:				label = "IEEE802154_nrf5";
dts/arm/ti/cc13x2_cc26x2.dtsi:				label = "IEEE802154_0";
dts/arm/ti/cc13x2_cc26x2.dtsi:				label = "IEEE802154_1";
```

gmarull committed:
dts: remove unused IEEE 802.15.4 labels

Labels are no longer needed because devices are obtaines using
zephyr,ieee802154 DT choice.

Signed-off-by: Gerard Marull-Paretas <gerard.marull@nordicsemi.no>

gmarull commented:
> Can you remove `label` usage from here as part of this PR:
> 
> ```
> dts/arm/nordic/nrf52811.dtsi:				label = "IEEE802154_nrf5";
> dts/arm/nordic/nrf52833.dtsi:				label = "IEEE802154_nrf5";
> dts/arm/nordic/nrf52840.dtsi:				label = "IEEE802154_nrf5";
> dts/arm/nordic/nrf5340_cpunet.dtsi:				label = "IEEE802154_nrf5";
> dts/arm/ti/cc13x2_cc26x2.dtsi:				label = "IEEE802154_0";
> dts/arm/ti/cc13x2_cc26x2.dtsi:				label = "IEEE802154_1";
> ```

done

gmarull commented:
> @gmarull what's up with bt510/bt610/nrf52840_blip that we remove `ieee802154` from yaml but no dts updates?

missed them, fixed thanks.

nandojve reviewed (approved):
Thank you @gmarull !
